flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/example/GSASingleSourceShortestPathsExample.java

```diff
@@ -0,0 +1,232 @@
+	//  Program
+	// --------------------------------------------------------------------------------------------
+
+	public static void main(String[] args) throws Exception {
+
+		if(!parseParameters(args)) {
+			return;
+		}
+
+		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+
+		DataSet<Vertex<Long, Double>> vertices = getVertexDataSet(env);
+		DataSet<Edge<Long, Double>> edges = getEdgeDataSet(env);
+
+		Graph<Long, Double, Double> graph = Graph.fromDataSet(vertices, edges, env);
+
+		DataSet<Vertex<Long, Double>> singleSourceShortestPaths = graph
+				.run(new GSASingleSourceShortestPathsExample()).getVertices();
+
+		// emit result
+		if(fileOutput) {
+			singleSourceShortestPaths.writeAsCsv(outputPath, "\n", " ");
+		} else {
+			singleSourceShortestPaths.print();
+		}
+
+		env.execute("GSA Single Source Shortest Paths Example");
+	}
+
+	// --------------------------------------------------------------------------------------------
+	//  GraphAlgorithm Functions
+	// --------------------------------------------------------------------------------------------
+
+	@Override
+	public Graph<Long, Double, Double> run(Graph<Long, Double, Double> input) {
+		// The path from src to trg through edge e costs src + e
+		// If the target's distance is 0 (the target is the actual source), return 0 instead
+		GatherFunction<Long, Double, Double, Double> gather = new SingleSourceShortestPathGather();
+
+		// Return the smaller path length to minimize distance
+		SumFunction<Long, Double, Double, Double> sum = new SingleSourceShortestPathSum();
+
+		// Iterate as long as the distance is updated
+		ApplyFunction<Long, Double, Double, Double> apply = new SingleSourceShortestPathApply();
+
+		GatherSumApplyIteration<Long, Double, Double, Double> iteration = input.createGatherSumApplyIteration(
+				gather, sum, apply, this.maxIterations);
+		return input.mapVertices(new InitVerticesMapper<Long>(this.srcVertexId))
+				.runGatherSumApplyIteration(iteration);
+	}
+
+	public static final class InitVerticesMapper<K extends Comparable<K> & Serializable>
+			implements MapFunction<Vertex<K, Double>, Double> {
+
+		private K srcVertexId;
+
+		public InitVerticesMapper(K srcId) {
+			this.srcVertexId = srcId;
+		}
+
+		public Double map(Vertex<K, Double> value) {
+			if (value.f0.equals(srcVertexId)) {
+				return 0.0;
+			} else {
+				return Double.POSITIVE_INFINITY;
+			}
+		}
+	}
+
+	// --------------------------------------------------------------------------------------------
+	//  Single Source Shortest Path UDFs
+	// --------------------------------------------------------------------------------------------
+
+	private static final class SingleSourceShortestPathGather
+			extends GatherFunction<Long, Double, Double, Double> {
+		@Override
+		public Tuple2<Long, Double> gather(Tuple3<Vertex<Long, Double>,
+				Edge<Long, Double>, Vertex<Long, Double>> triplet) {
```

Comment:
How about we make a custom `Triplet` type and add `getSrcVertex()`, `getEdge()` and `getTrgVertex()` methods, so that we don't need to use the tuple fields here? I think it makes code a bit hard to understand..